Automatically add references to core framework assemblies when targeting .NET Framework

dsplaisted opened the pull request:
Right now, if you create an ASP.NET Core application targeting .NET Framework, you won't get references to any framework assemblies except mscorlib.  This means a bunch of types (`Uri`, Linq, etc) won't be available to you.

We could add these to the templates, but in the interest of keeping the `.csproj` files concise I think it's better to reference them automatically in the SDK targets.  That's what this PR does.

The list of assemblies comes from what you get by default when creating a new .NET Framework Class Library.  For parity with .NET Standard and .NET Core projects, I'd suggest we should reference all assemblies that contain types that are in .NET Standard 2.0.  @weshaggard @ericstj @terrajobst can you provide a list of assemblies you would need to reference in .NET Framework to get all the types that are in .NET Standard 2.0?

dsplaisted committed:
Automatically add references to core framework assemblies when targeting .NET Framework

davidfowl commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -65,6 +65,26 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <_TargetFrameworkMinorVersion Condition="$(_TargetFrameworkVersionSegmentCount) > 1">$(TargetFrameworkVersion.Split('.')[1])</_TargetFrameworkMinorVersion>
+  </PropertyGroup>
+  
+  <ItemGroup Condition=" '$(DisableImplicitFrameworkReferences)' != 'true' and '$(TargetFrameworkIdentifier)' == '.NETFramework'">
```
This is bigger than the original set of default references in project.json/DNX. How were these chosen? Based on the .NET Framework default template today? Can we make this a bit smaller?

dsplaisted replied:
> Based on the .NET Framework default template today?

Yes, that's what I based it on.  However, I'm suggesting that we make it even bigger by default, to include all the types that are in .NET Standard 2.0.  That way you'll have access to all .NET Standard 2.0 types by default whether you are targeting .NET Core or .NET Framework.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -65,6 +65,26 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <_TargetFrameworkMajorVersion>$(TargetFrameworkVersion.Split('.')[0])</_TargetFrameworkMajorVersion>
+    <_TargetFrameworkMajorVersion Condition="$(_TargetFrameworkMajorVersion.StartsWith('v'))">$(TargetFrameworkVersion.Substring(1))</_TargetFrameworkMajorVersion>
+    <_TargetFrameworkMinorVersion Condition="$(_TargetFrameworkVersionSegmentCount) > 1">$(TargetFrameworkVersion.Split('.')[1])</_TargetFrameworkMinorVersion>
+  </PropertyGroup>
```
I don't think this string manipulation is necessary. IIRC, @rainersigwald mentioned elsewhere that MSBuild can will automagically compare System.Version-parseable strings and compare them as such.

rainersigwald replied:
That is true, though `System.Version.TryParse` will fail on a `v`-prefixed string so you'll probably need to do that still.

dsplaisted replied:
@rainersigwald What's the syntax to compare version numbers in MSBuild?  Or do strings that parse as version numbers automatically get compared as such?

rainersigwald replied:
The latter: strings that parse as version numbers get compared that way with the usual comparison operators.

nguerrera reviewed (changes requested):
Migration currently adds the default refs from project.json. If we land on a superset of that, then we should adjust the migration. File a bug on CLI for that.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -65,6 +65,26 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <Reference Include="Microsoft.CSharp"/>
+    <Reference Include="System.Data"/>
+    <Reference Include="System.Net.Http" Condition="($(_TargetFrameworkMajorVersion) > 4) Or ($(_TargetFrameworkMajorVersion) == 4 And $(_TargetFrameworkMinorVersion) >= 5)"/>
+    <Reference Include="System.Xml"/>
```
nit: sort these alphabetically.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -65,6 +65,26 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <Reference Include="System.Data.DataSetExtensions"/>
+    <Reference Include="Microsoft.CSharp"/>
+    <Reference Include="System.Data"/>
+    <Reference Include="System.Net.Http" Condition="($(_TargetFrameworkMajorVersion) > 4) Or ($(_TargetFrameworkMajorVersion) == 4 And $(_TargetFrameworkMinorVersion) >= 5)"/>
```
There are other conditions needed for version < 4, right?

nguerrera replied:
Tests won't pass with version < 4 but that's only because of microsoft/msbuild#1333 combined with tests always running with core msbuild as of now. We keep getting requests for it, so we shouldn't discount version < 4 here.

dsplaisted committed:
Add the rest of default references that have .NET Standard types

Also update the test to make sure there aren't any warnings about references that can't be found

dsplaisted commented:
I've updated the list of references to be based on what is necessary to make the types in .NET Standard available, based on this list: https://github.com/dotnet/standard/tree/master/platforms/net461

From that list, I've excluded `System.ComponentModel.Composition`, which is only there to support a type forward for `Lazy<T, TMetadata>`, and `System.Web`, which is only there for `HttpUtility`, but has a whole bunch of other stuff we don't want to bring in.

I've also added conditions to the references so that warnings won't be generated when you target .NET Framework 3.5 or 4.0, which don't have all of these assemblies.  I've also sorted them.

dsplaisted commented:
@nguerrera This is ready for you to re-review now

nguerrera commented:
Also, is there a bug on the CLI to remove now redundant references from migration. What will be the behavior for migrated projects from the current preview when they get the next one and start to have dup references?

dsplaisted commented:
> Also, is there a bug on the CLI to remove now redundant references from migration. What will be the behavior for migrated projects from the current preview when they get the next one and start to have dup references?

I filed dotnet/cli#4695 to remove the redundant references.  Duplicate references don't appear to be a problem, I think MSBuild ends up merging them.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -65,6 +65,27 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <Reference Include="System.Numerics" Condition=" '$(_TargetFrameworkVersionWithoutV)' >= '4.0' "/>
+    <Reference Include="System.Runtime.Serialization"/>
+    <Reference Include="System.Xml.Linq"/>
+    <Reference Include="System.Xml"/>
```
I don't have a strong opinion, but I left it out because it's not currently included in .NET Standard.

dsplaisted committed:
Add clarifying comment
